File: test/cpu/or.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"
#include "inline.hpp"
#include "test.hpp"
#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
struct F_OR {
    F_OR() {
        // OR.B %D2, %D3
        TEST::SET_W(0, 0100000 | 3 << 9 | 2);
        TEST::SET_W(2, TEST_BREAK);

        // OR.W %D2, %D3
        TEST::SET_W(4, 0100100 | 3 << 9 | 2);
        TEST::SET_W(6, TEST_BREAK);

        // OR.L %D2, %D3
        TEST::SET_W(8, 0100200 | 3 << 9 | 2);
        TEST::SET_W(10, TEST_BREAK);

        // OR.B %D2, (%A3)
        TEST::SET_W(12, 0100420 | 3 << 9 | 2);
        TEST::SET_W(14, TEST_BREAK);

        // OR.W %D2, (%A3)
        TEST::SET_W(16, 0100520 | 3 << 9 | 2);
        TEST::SET_W(18, TEST_BREAK);

        // OR.L %D2, (%A3)
        TEST::SET_W(20, 0100620 | 3 << 9 | 2);
        TEST::SET_W(22, TEST_BREAK);
        jit_compile(0, 24);
    }};
BOOST_FIXTURE_TEST_SUITE(OR, Prepare, *boost::unit_test::fixture<F_OR>())


BOOST_AUTO_TEST_SUITE(Byte)
BOOST_AUTO_TEST_SUITE(ToReg)
BOOST_AUTO_TEST_CASE(value) {
    cpu.D[3] = 0x30;
    cpu.D[2] = 0x2F;
    run_test(0);
    BOOST_TEST(cpu.D[3] == 0x3F);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(n) {
    cpu.D[3] = 0x80;
    cpu.D[2] = 0;
    run_test(0);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(z) {
    cpu.D[3] = 0;
    cpu.D[2] = 0;
    run_test(0);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(ToMem)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[3] = 0x2F;
    cpu.A[2] = 0x1000;
    RAM[0x1000] = 0x30;
    run_test(12);
    BOOST_TEST(RAM[0x1000] == 0x3F);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(n) {
    cpu.D[3] = 0;
    cpu.A[2] = 0x1000;
    RAM[0x1000] = 0x80;
    run_test(12);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(z) {
    cpu.D[3] = 0;
    cpu.A[2] = 0x1000;
    RAM[0x1000] = 0;
    run_test(12);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Word)
BOOST_AUTO_TEST_SUITE(ToReg)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[3] = 0x3030;
    cpu.D[2] = 0x2f2f;
    run_test(4);
    BOOST_TEST(cpu.D[3] == 0x3F3F);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}
BOOST_AUTO_TEST_CASE(n) {
    cpu.D[3] = 0x8000;
    cpu.D[2] = 0;
    run_test(4);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(z) {
    cpu.D[3] = 0;
    cpu.D[2] = 0;
    run_test(4);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(ToMem)
BOOST_AUTO_TEST_CASE(value) {
    cpu.D[3] = 0x2F2F;
    TEST::SET_W(0x1000, 0x3030);
    cpu.A[2] = 0x1000;
    run_test(16);
    BOOST_TEST(TEST::GET_W(0x1000) == 0x3F3F);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(n) {
    cpu.D[3] = 0;
    TEST::SET_W(0x1000, 0x8000);
    cpu.A[2] = 0x1000;
    run_test(16);
    BOOST_TEST(cpu.N);
}
BOOST_AUTO_TEST_CASE(z) {
    cpu.D[3] = 0;
    TEST::SET_W(0x1000, 0);
    cpu.A[2] = 0x1000;
    run_test(16);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(Long)
BOOST_AUTO_TEST_SUITE(ToReg)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[3] = 0x30303030;
    cpu.D[2] = 0x2f2f2f2f;
    run_test(8);
    BOOST_TEST(cpu.D[3] == 0x3f3f3f3f);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}
BOOST_AUTO_TEST_CASE(n) {
    cpu.D[3] = 0x80000000;
    cpu.D[2] = 0;
    run_test(8);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(z) {
    cpu.D[3] = 0;
    cpu.D[2] = 0;
    run_test(8);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(ToMem)
BOOST_AUTO_TEST_CASE(value) {
    cpu.D[3] = 0x2F2F2F2F;
    TEST::SET_L(0x1000, 0x30303030);
    cpu.A[2] = 0x1000;
    run_test(20);
    BOOST_TEST(TEST::GET_L(0x1000) == 0x3F3F3F3F);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}
BOOST_AUTO_TEST_CASE(n) {
    cpu.D[3] = 0;
    TEST::SET_L(0x1000, 0x80000000);
    cpu.A[2] = 0x1000;
    run_test(20);
    BOOST_TEST(cpu.N);
}
BOOST_AUTO_TEST_CASE(z) {
    cpu.D[3] = 0;
    TEST::SET_L(0x1000, 0);
    cpu.A[2] = 0x1000;
    run_test(20);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/or.lua
SRC = math.random(1, 8)
repeat
   DST = math.random(1, 8)
until DST ~= SRC

ADDR = math.random(1, 7) -- don't use SP

-- OR.B %Dx, %Dy
MEM[0].W = OCT"0100000" | ((DST-1) << 9) | (SRC-1)
MEM[2].W = TEST.BREAK

-- OR.W %Dx, %Dy
MEM[4].W = OCT"0100100" | ((DST-1) << 9) | (SRC-1)
MEM[6].W = TEST.BREAK

-- OR.L %Dx, %Dy
MEM[8].W = OCT"0100200" | ((DST-1) << 9) | (SRC-1)
MEM[10].W = TEST.BREAK

-- OR.B %Dx, (%Az)
MEM[12].W = OCT"0100420" | ((SRC-1) << 9) | (ADDR-1)
MEM[14].W = TEST.BREAK

-- OR.W %Dx, (%Az)
MEM[16].W = OCT"0100520" | ((SRC-1) << 9) | (ADDR-1)
MEM[18].W = TEST.BREAK

-- OR.L %Dx, (%Az)
MEM[20].W = OCT"0100620" | ((SRC-1) << 9) | (ADDR-1)
MEM[22].W = TEST.BREAK

JIT_COMPILE(0, 24)

run = { Byte = {ToReg = {}, ToMem = {}},
		Word = {ToReg = {}, ToMem = {}},
		Long = {ToReg = {}, ToMem = {}},
	  }
function run.Byte.ToReg.value()
   CPU.D[DST].B = 0x30
   CPU.D[SRC].B = 0x2F
   TEST.RUN(0)

   TEST.CHECK ("CPU.D[DST].B", 0x3F)
   TEST.CHECK ("CPU.CC.V", false)
   TEST.CHECK ("CPU.CC.C", false)
   TEST.CHECK ("CPU.CC.N", false)
   TEST.CHECK ("CPU.CC.Z", false)
end
function run.Byte.ToReg.n()
   CPU.D[DST].B = 0x80
   CPU.D[SRC].B = 0
   TEST.RUN(0)
   TEST.CHECK ("CPU.CC.N", true)
end

function run.Byte.ToReg.z()
   CPU.D[DST].B = 0
   CPU.D[SRC].B = 0
   TEST.RUN(0)
   TEST.CHECK ("CPU.CC.Z", true)
end

function run.Byte.ToMem.value()
   CPU.D[SRC].B = 0x2F
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].B = 0x30
   TEST.RUN(12)
   TEST.CHECK ("MEM[0x1000].B", 0x3F)
   TEST.CHECK ("CPU.CC.V", false)
   TEST.CHECK ("CPU.CC.C", false)
   TEST.CHECK ("CPU.CC.N", false)
   TEST.CHECK ("CPU.CC.Z", false)
end

function run.Byte.ToMem.n()
   CPU.D[SRC].B = 0
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].B = 0x80
   TEST.RUN(12)
   TEST.CHECK( "CPU.CC.N", true)
end

function run.Byte.ToMem.z()
   CPU.D[SRC].B = 0
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].B = 0
   TEST.RUN(12)
   TEST.CHECK( "CPU.CC.Z", true)
end

function run.Word.ToReg.value()
   CPU.D[DST].W = 0x3030
   CPU.D[SRC].W = 0x2F2F
   TEST.RUN(4)

   TEST.CHECK ("CPU.D[DST].W", 0x3F3F)
   TEST.CHECK ("CPU.CC.V", false)
   TEST.CHECK ("CPU.CC.C", false)
   TEST.CHECK ("CPU.CC.N", false)
   TEST.CHECK ("CPU.CC.Z", false)
end

function run.Word.ToReg.n()
   CPU.D[DST].W = 0x8000
   CPU.D[SRC].W = 0
   TEST.RUN(4)
   TEST.CHECK ("CPU.CC.N", true)
end

function run.Word.ToReg.z()
   CPU.D[DST].W = 0
   CPU.D[SRC].W = 0
   TEST.RUN(4)
   TEST.CHECK ("CPU.CC.Z", true)
end

function run.Word.ToMem.value()
   CPU.D[SRC].W = 0x2F2F
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].W = 0x3030
   TEST.RUN(16)
   TEST.CHECK ("MEM[0x1000].W", 0x3F3F)
   TEST.CHECK ("CPU.CC.V", false)
   TEST.CHECK ("CPU.CC.C", false)
   TEST.CHECK ("CPU.CC.N", false)
   TEST.CHECK ("CPU.CC.Z", false)
end

function run.Word.ToMem.n()
   CPU.D[SRC].W = 0
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].W = 0x8000
   TEST.RUN(16)
   TEST.CHECK( "CPU.CC.N", true)
end

function run.Word.ToMem.z()
   CPU.D[SRC].W = 0
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].W = 0
   TEST.RUN(16)
   TEST.CHECK( "CPU.CC.Z", true)
end

function run.Long.ToReg.value()
   CPU.D[DST].L = 0x30303030
   CPU.D[SRC].L = 0x2F2F2F2F
   TEST.RUN(8)

   TEST.CHECK ("CPU.D[DST].L", 0x3F3F3F3F)
   TEST.CHECK ("CPU.CC.V", false)
   TEST.CHECK ("CPU.CC.C", false)
   TEST.CHECK ("CPU.CC.N", false)
   TEST.CHECK ("CPU.CC.Z", false)
end
function run.Long.ToReg.n()
   CPU.D[DST].L = 0x80000000
   CPU.D[SRC].L = 0
   TEST.RUN(8)
   TEST.CHECK ("CPU.CC.N", true)
end

function run.Long.ToReg.z()
   CPU.D[DST].L = 0
   CPU.D[SRC].L = 0
   TEST.RUN(8)
   TEST.CHECK ("CPU.CC.Z", true)
end

function run.Long.ToMem.value()
   CPU.D[SRC].L = 0x2F2F2F2F
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].L = 0x30303030
   TEST.RUN(20)
   TEST.CHECK ("MEM[0x1000].L", 0x3F3F3F3F)
   TEST.CHECK ("CPU.CC.V", false)
   TEST.CHECK ("CPU.CC.C", false)
   TEST.CHECK ("CPU.CC.N", false)
   TEST.CHECK ("CPU.CC.Z", false)
end

function run.Long.ToMem.n()
   CPU.D[SRC].L = 0
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].L = 0x80000000
   TEST.RUN(20)
   TEST.CHECK( "CPU.CC.N", true)
end

function run.Long.ToMem.z()
   CPU.D[SRC].L = 0
   CPU.A[ADDR].L = 0x1000
   MEM[0x1000].L = 0
   TEST.RUN(20)
   TEST.CHECK( "CPU.CC.Z", true)
end
